Feat/simulation

AzimovS opened the pull request:
In this PR:
1. Should fix #27. The table refreshed every 5 seconds.
2. Feat: #25, now it is possible to edit the price by clicking the cell with the price.
<img width="932" height="566" alt="image" src="https://github.com/user-attachments/assets/90067d80-d30f-4352-a44f-4782916ad728" />

3. Added a simulation toggle #29.

<img width="958" height="587" alt="image" src="https://github.com/user-attachments/assets/c3cf4acd-37d6-406d-b7e7-0fff68c1dabc" />


To implement the simulation toggle:
-  I used wagmi write contract instead of scaffold, because I didn't want to spam with a lot of notification "Transaction completed successfully!". Also, I didn't want to create writeContractAsync for each SimpleOracle (and we have a lot of them).
- Between transactions I added 200 ms delay, otherwise there was an issue with the nonce


Another important issue we need to discuss is the last reported time. In Hardhat, the blockchain time runs ahead of real time because we are sending a large number of transactions. For example, when we deploy the contracts, the last reported time is already ahead of the actual time. Here are some possible solutions that might work:
-- Set automining to false and use an interval of 1 second.
-- Display the blockchain time as well, so students can rely on it instead of the system clock.
-- Move the simulation script to a terminal script. Temporarily turn off automining, mine all transactions using the current time, and then turn it back on. The same approach should be applied to the deployment script, but maybe if the user clicks a lot of times faucet button, the time will be desync.
-- The time is important for the active/stale reporters, maybe we can remove this feature for simplicity, and mention that the reported time is ahead of the real time


I will be glad for any suggestion

escottalexander reviewed (changes requested):
When I run this I see duplicate entries on the front-end. I deployed twice (maybe related, though I couldn't reproduce) but it seems strange that there would be duplicate rows with the same address.
<img width="996" height="988" alt="image" src="https://github.com/user-attachments/assets/731cd3df-4375-4985-a777-eee26837100f" />

I get this error when I try to toggle simulation. Did I miss something? I am using a burner wallet. Not sure if that would be a problem.
<img width="825" height="285" alt="image" src="https://github.com/user-attachments/assets/eb7aac75-2cb5-4636-b127-d3f0211692ab" />
Later, I get this error when trying to toggle it:
<img width="731" height="220" alt="image" src="https://github.com/user-attachments/assets/7476d57d-10e0-4b07-ada7-7f5cbbd5b54d" />

I think we should add a pencil icon next to the prices so that it is obvious they can be edited.

Let me know if you need to debug anything on my end. Weird that I am facing those errors.

escottalexander commented:
> -- The time is important for the active/stale reporters, maybe we can remove this feature for simplicity, and mention that the reported time is ahead of the real time

Perhaps it would be better to not show times at all and instead show something like:
- "< 10s"  (or whatever the threshold is for stale) or "recent"
- "stale" (could even be greyed out)
Totally open to thoughts on that.
Otherwise users are having to compare timestamps to the actual time (and quickly!) so this may be a lot easier to understand.

AzimovS commented:
Hi!

I rebased on #35 , sorry for the order.


> When I run this I see duplicate entries on the front-end. I deployed twice (maybe related, though I couldn't reproduce) but it seems strange that there would be duplicate rows with the same address.

After `yarn deploy` hardhat is reusing the contract and in our contract we can add the same oracle. Not sure if it is a bug, but maybe it is. Probably deploying with `yarn deploy --reset` should fix that. 

> I get this error when I try to toggle simulation. Did I miss something? I am using a burner wallet. Not sure if that would be a problem.
Fixed errors with the simulation when there are no valid prices. 




> I think we should add a pencil icon next to the prices so that it is obvious they can be edited.

refactored, thanks!

> Perhaps it would be better to not show times at all and instead show something like:
> 
> * "< 10s"  (or whatever the threshold is for stale) or "recent" or even a count-up "1s ago" ... "2s ago"
> * "stale"
>   Totally open to thoughts on that.
>   Otherwise users are having to compare timestamps to the actual time (and quickly!) so this may be a lot easier to understand.

I tried to show "1s ago". It means we need to fetch the latest time from the blockchain, but if there are no transactions, then it will keep just fetching the same timestamp :(

AzimovS commented:
The simulation part was transferred to the backend. I had an error that when I stopped the script, it didn't turn on the automine in the end.  I added cleanup on the script termintation for both scripts.

Also, I removed triggering txs on the frontend for non burner wallets. Otherwise, it is weird getting pop up window to sign the tx every several seconds
